### ISSUE TITLE
Custom MLIR Tensorflow Pass

### ISSUE THREAD

**Issue by nicholaiTukanov** [tensorflow/tensorflow#55750]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Feature Request

### Source

source

### Tensorflow Version

2.8.0

### Custom Code

Yes

### OS Platform and Distribution

Linux Centos 7

### Python version

3.10.4

### Bazel version

4.2.1

### GCC/Compiler version

11.2.1

### Current Behaviour?

Hi, I am currently working a compiler project that identifies when complex arithmetic is being cast as real arithmetic. For example, say we want to compute 
```
C = tf.matmul(A, B)
```
where A, B, C are tf.complex64. But, due to whatever reason it is not allowed, or the person writing is a bad developer. Thus, instead we use real arithmetic to compute 
```
Cr = tf.matmul(Ar, Br) - tf.matmul(Ai, Bi)
Ci = tf.matmul(Ar, Bi) + tf.matmul(Ai, Br)
```
where Ar, Ai, Br, Bi, Cr, Ci are tf.float32. I would like to catch this in a compiler pass and modify the program to use a high performance complex matmul kernel instead. As well as convert the data format into the one expected by the kernel.

I was looking into the [Tensorflow MLIR passes](https://www.tensorflow.org/mlir/tf_passes) and was wondering how could I go about writing my own pass to implement this. So far I have been trying to convert `tf` MLIR to `linalg` MLIR and then writing an `OperationPass` for this. It would be a lot easier if I could just add a custom pass such that I can use it within `tf-opt`

TLDR: Are there any docs on writing custom [tensorflow passes](https://www.tensorflow.org/mlir/tf_passes)? If not, can these be added?

### Standalone code to reproduce the issue

```
# Assumes tf 2.8
import tensorflow as tf

@tf.function
def cgemm_explict(A_r, A_i, B_r, B_i, C_r, C_i):
    return tf.complex(A_r @ B_r - A_i @ B_i + C_r, A_r @ B_i + A_i @ B_r + C_i)

m, n, k = 3, 3, 3
func_obj = cgemm_explict.get_concrete_function(
    tf.TensorSpec(shape=[m, k], dtype=tf.dtypes.float32, name="Ar"),
    tf.TensorSpec(shape=[m, k], dtype=tf.dtypes.float32, name="Ai"),
    tf.TensorSpec(shape=[k, n], dtype=tf.dtypes.float32, name="Br"),
    tf.TensorSpec(shape=[k, n], dtype=tf.dtypes.float32, name="Bi"),
    tf.TensorSpec(shape=[m, n], dtype=tf.dtypes.float32, name="Cr"),
    tf.TensorSpec(shape=[m, n], dtype=tf.dtypes.float32, name="Ci")
)
# replace 'tf-standard-pipeline' with custom pass
mlir_obj = tf.mlir.experimental.convert_function(func_obj, pass_pipeline='tf-standard-pipeline')

print(mlir_obj) # print tf MLIR for the given tf.function
```

### Relevant log output

```shell
module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, producer = 987 : i32}}  {
  func @__inference_cgemm_explict_19(%arg0: tensor<3x3xf32> {tf._user_specified_name = "Ar"}, %arg1: tensor<3x3xf32> {tf._user_specified_name = "Ai"}, %arg2: tensor<3x3xf32> {tf._user_specified_name = "Br"}, %arg3: tensor<3x3xf32> {tf._user_specified_name = "Bi"}, %arg4: tensor<3x3xf32> {tf._user_specified_name = "Cr"}, %arg5: tensor<3x3xf32> {tf._user_specified_name = "Ci"}) -> tensor<3x3xcomplex<f32>> attributes {tf.entry_function = {control_outputs = "", inputs = "ar,ai,br,bi,cr,ci", outputs = "identity_RetVal"}} {
    %0 = "tf.MatMul"(%arg1, %arg3) {device = "", transpose_a = false, transpose_b = false} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %1 = "tf.MatMul"(%arg0, %arg3) {device = "", transpose_a = false, transpose_b = false} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %2 = "tf.MatMul"(%arg0, %arg2) {device = "", transpose_a = false, transpose_b = false} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %3 = "tf.Sub"(%2, %0) {device = ""} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %4 = "tf.MatMul"(%arg1, %arg2) {device = "", transpose_a = false, transpose_b = false} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %5 = "tf.AddV2"(%1, %4) {device = ""} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %6 = "tf.AddV2"(%5, %arg5) {device = ""} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %7 = "tf.AddV2"(%3, %arg4) {device = ""} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xf32>
    %8 = "tf.Complex"(%7, %6) {device = ""} : (tensor<3x3xf32>, tensor<3x3xf32>) -> tensor<3x3xcomplex<f32>>
    %9 = "tf.Identity"(%8) {device = ""} : (tensor<3x3xcomplex<f32>>) -> tensor<3x3xcomplex<f32>>
    return %9 : tensor<3x3xcomplex<f32>>
  }
}
```
</details>

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in tf v2.7, v2.8 and nightly.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/b94c80e9825757fa05ddb77b0281deff/55750.ipynb).

**Comment by jpienaar** [comment]
Good question and quite timely with the new function rewrite API being worked on. Could you clarify if you are interested in having this just in offline tool (a la tf-opt or tfg_graph_transforms)?

**Comment by nicholaiTukanov** [comment]
An offline tool would be great, since it makes it easier for me to test and play around with. 

When thinking for my project, I thought it would be something like tf-opt. But, I wasn't aware of tfg_graph_transforms. Are you referring to the [Graph Transform Tool](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#writing-your-own-transforms)?

**Comment by jpienaar** [comment]
No, https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/tfg_graph_transforms (which uses TFG dialect which is more isomorphic with GraphDef than TF). But there isn't guide to either unfortunately, but depending on your use it's easier to point to different folks and ask. The passes themselves are standard MLIR ones and you seem like you know that part (the ops I can point you to their definitions in either dialect too). 

tf-opt is more of a testing tool, so the UI is more basic/sharp edges but it's fairly simple to add a pass. There isn't a registration mechanism to add a pass beyond just modifying it to register your pass and changing the BUILD file (alternatively one can just create a new foo-opt tool and register the dialects and passes one want/copy those from tf-opt and add new ones, this should be possible wherever you can add bazel dep). But it mostly works well enough for experimentation/playing.

For the other one I could have sworn we had some mechanism recently, but if not it would work same as above.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
